File: android/src/test/java/de/idrinth/habitevaluator/android/ui/AdapterDataConsistencyTest.java
package de.idrinth.habitevaluator.android.ui;

import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

import de.idrinth.habitevaluator.shared.model.DiaryEntry;
import de.idrinth.habitevaluator.shared.model.EmotionPair;
import de.idrinth.habitevaluator.shared.model.EventSignificance;
import de.idrinth.habitevaluator.shared.model.FoodLog;
import de.idrinth.habitevaluator.shared.model.Habit;
import de.idrinth.habitevaluator.shared.model.HabitCategory;
import de.idrinth.habitevaluator.shared.model.Medication;
import de.idrinth.habitevaluator.shared.model.MedicationLog;
import de.idrinth.habitevaluator.shared.model.ActivityLog;
import de.idrinth.habitevaluator.shared.model.SleepEntry;
import de.idrinth.habitevaluator.shared.model.SportLog;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Regression test: every item in the adapter's backing data must be represented
 * in getItemCount(). If getItemCount() returns fewer items than the data holds,
 * the RecyclerView will silently drop rows and the user cannot scroll to them.
 */
class AdapterDataConsistencyTest {

    // ── HabitAdapter ────────────────────────────────────────────────────────

    @Test
    void testHabitAdapterItemCountAlwaysMatchesBackingList() {
        List<Habit> habits = new ArrayList<>();
        HabitAdapter adapter = new HabitAdapter(habits, new NoOpHabitClickListener());

        for (int i = 0; i < 50; i++) {
            habits.add(new Habit("Habit " + i, "desc"));
            assertEquals(
                    habits.size(), adapter.getItemCount()
            ,
                    "After adding item " + i + ", getItemCount must equal list size");
        }

        while (!habits.isEmpty()) {
            habits.remove(habits.size() - 1);
            assertEquals(
                    habits.size(), adapter.getItemCount()
            ,
                    "After removing, getItemCount must equal list size");
        }
    }

    @Test
    void testHabitAdapterItemCountAfterBulkClear() {
        List<Habit> habits = new ArrayList<>();
        HabitAdapter adapter = new HabitAdapter(habits, new NoOpHabitClickListener());

        for (int i = 0; i < 20; i++) {
            habits.add(new Habit("H" + i, "d"));
        }
        assertEquals(20, adapter.getItemCount());

        habits.clear();
        assertEquals(0, adapter.getItemCount());
    }

    // ── DiaryEntryAdapter ───────────────────────────────────────────────────

    @Test
    void testDiaryEntryAdapterItemCountAlwaysMatchesBackingList() {
        List<DiaryEntry> entries = new ArrayList<>();
        DiaryEntryAdapter adapter = new DiaryEntryAdapter(entries, null);

        for (int i = 0; i < 50; i++) {
            DiaryEntry entry = new DiaryEntry();
            entry.setDescription("Entry " + i);
            entry.setSignificance(EventSignificance.NORMAL);
            entry.setEventDate(LocalDate.now());
            entries.add(entry);
            assertEquals(entries.size(), adapter.getItemCount());
        }

        while (!entries.isEmpty()) {
            entries.remove(entries.size() - 1);
            assertEquals(entries.size(), adapter.getItemCount());
        }
    }

    // ── SleepEntryAdapter ───────────────────────────────────────────────────

    @Test
    void testSleepEntryAdapterItemCountAlwaysMatchesBackingList() {
        List<SleepEntry> entries = new ArrayList<>();
        SleepEntryAdapter adapter = new SleepEntryAdapter(entries, null);

        for (int i = 0; i < 50; i++) {
            SleepEntry entry = new SleepEntry();
            entry.setDate(LocalDate.now().minusDays(i));
            entry.setFromTime(LocalTime.of(23, 0));
            entry.setUntilTime(LocalTime.of(7, 0));
            entries.add(entry);
            assertEquals(entries.size(), adapter.getItemCount());
        }

        while (!entries.isEmpty()) {
            entries.remove(entries.size() - 1);
            assertEquals(entries.size(), adapter.getItemCount());
        }
    }

    // ── EmotionPairAdapter ──────────────────────────────────────────────────

    @Test
    void testEmotionPairAdapterItemCountAlwaysMatchesBackingList() {
        List<EmotionPair> pairs = new ArrayList<>();
        EmotionPairAdapter adapter = new EmotionPairAdapter(pairs, null, null);

        for (int i = 0; i < 30; i++) {
            EmotionPair pair = new EmotionPair();
            pair.setNegativeLabel("neg" + i);
            pair.setPositiveLabel("pos" + i);
            pairs.add(pair);
            assertEquals(pairs.size(), adapter.getItemCount());
        }

        while (!pairs.isEmpty()) {
            pairs.remove(pairs.size() - 1);
            assertEquals(pairs.size(), adapter.getItemCount());
        }
    }

    // ── EmotionDataAdapter (expandable groups) ──────────────────────────────
    // Note: EmotionDataAdapter.setData() calls notifyDataSetChanged() which
    // requires the Android RecyclerView runtime. These tests verify the adapter's
    // initial empty state — the contract that getItemCount() == 0 before any
    // data is loaded, preventing IndexOutOfBoundsException on first layout pass.

    @Test
    void testEmotionDataAdapterStartsEmpty() {
        EmotionDataAdapter adapter = new EmotionDataAdapter(pair -> {}, pair -> {}, entry -> {});
        assertEquals(
                0, adapter.getItemCount()
        ,
                "EmotionDataAdapter must start with zero items before setData is called");
    }

    @Test
    void testEmotionDataAdapterWithNullListenersStartsEmpty() {
        EmotionDataAdapter adapter = new EmotionDataAdapter(null, null, null);
        assertEquals(0, adapter.getItemCount());
    }

    // ── EditHabitAdapter ────────────────────────────────────────────────────

    @Test
    void testEditHabitAdapterItemCountMatchesBackingList() {
        List<Habit> habits = new ArrayList<>();
        List<HabitCategory> categories = new ArrayList<>();
        HabitCategory cat = new HabitCategory("Health", "Health category", "#00FF00");
        categories.add(cat);

        for (int i = 0; i < 20; i++) {
            Habit h = new Habit("Habit " + i, "desc " + i);
            h.setCategoryId(cat.getId());
            habits.add(h);
        }

        EditHabitAdapter adapter = new EditHabitAdapter(habits, false, categories, "en");
        assertEquals(habits.size(), adapter.getItemCount());
    }

    @Test
    void testEditHabitAdapterEmptyList() {
        List<Habit> habits = new ArrayList<>();
        List<HabitCategory> categories = new ArrayList<>();
        EditHabitAdapter adapter = new EditHabitAdapter(habits, false, categories, "en");
        assertEquals(0, adapter.getItemCount());
    }

    // ── SportLogAdapter ─────────────────────────────────────────────────────

    @Test
    void testSportLogAdapterItemCountAlwaysMatchesBackingList() {
        List<SportLog> entries = new ArrayList<>();
        SportLogAdapter adapter = new SportLogAdapter(entries, null);

        for (int i = 0; i < 30; i++) {
            entries.add(new SportLog("Running", 5.0, "km",
                    LocalTime.of(8, 0), LocalTime.of(9, 0)));
            assertEquals(entries.size(), adapter.getItemCount());
        }

        entries.clear();
        assertEquals(0, adapter.getItemCount());
    }

    // ── FoodLogAdapter ──────────────────────────────────────────────────────

    @Test
    void testFoodLogAdapterItemCountAlwaysMatchesBackingList() {
        List<FoodLog> entries = new ArrayList<>();
        FoodLogAdapter adapter = new FoodLogAdapter(entries, null);

        for (int i = 0; i < 30; i++) {
            entries.add(new FoodLog(10.0, 200, LocalDateTime.now(), "Food " + i));
            assertEquals(entries.size(), adapter.getItemCount());
        }

        entries.clear();
        assertEquals(0, adapter.getItemCount());
    }

    // ── MedicationLogAdapter ────────────────────────────────────────────────

    @Test
    void testMedicationLogAdapterItemCountAlwaysMatchesBackingList() {
        List<MedicationLog> entries = new ArrayList<>();
        MedicationLogAdapter adapter = new MedicationLogAdapter(entries, null);

        for (int i = 0; i < 30; i++) {
            entries.add(new MedicationLog(null, 1.0, LocalDateTime.now()));
            assertEquals(entries.size(), adapter.getItemCount());
        }

        entries.clear();
        assertEquals(0, adapter.getItemCount());
    }

    // ── ActivityLogAdapter ────────────────────────────────────────────────────

    @Test
    void testActivityLogAdapterItemCountAlwaysMatchesBackingList() {
        List<ActivityLog> entries = new ArrayList<>();
        ActivityLogAdapter adapter = new ActivityLogAdapter(entries, null);

        for (int i = 0; i < 30; i++) {
            entries.add(new ActivityLog("Person " + i, "Location " + i,
                    LocalTime.of(10, 0), LocalTime.of(11, 0)));
            assertEquals(entries.size(), adapter.getItemCount());
        }

        entries.clear();
        assertEquals(0, adapter.getItemCount());
    }

    // ── Cross-adapter: bulk insert + remove never desynchronizes ────────────

    @Test
    void testAllSimpleAdaptersStaySynchronizedDuringMixedOperations() {
        // HabitAdapter
        List<Habit> habits = new ArrayList<>();
        HabitAdapter habitAdapter = new HabitAdapter(habits, new NoOpHabitClickListener());

        // SleepEntryAdapter
        List<SleepEntry> sleeps = new ArrayList<>();
        SleepEntryAdapter sleepAdapter = new SleepEntryAdapter(sleeps, null);

        // DiaryEntryAdapter
        List<DiaryEntry> diaries = new ArrayList<>();
        DiaryEntryAdapter diaryAdapter = new DiaryEntryAdapter(diaries, null);

        // Add items to all three
        for (int i = 0; i < 10; i++) {
            habits.add(new Habit("H" + i, "d"));
            SleepEntry se = new SleepEntry();
            se.setDate(LocalDate.now().minusDays(i));
            se.setFromTime(LocalTime.of(23, 0));
            se.setUntilTime(LocalTime.of(7, 0));
            sleeps.add(se);
            DiaryEntry de = new DiaryEntry();
            de.setDescription("D" + i);
            de.setSignificance(EventSignificance.MINOR);
            de.setEventDate(LocalDate.now());
            diaries.add(de);
        }

        assertEquals(10, habitAdapter.getItemCount());
        assertEquals(10, sleepAdapter.getItemCount());
        assertEquals(10, diaryAdapter.getItemCount());

        // Remove every other item
        for (int i = 9; i >= 0; i -= 2) {
            habits.remove(i);
            sleeps.remove(i);
            diaries.remove(i);
        }

        assertEquals(habits.size(), habitAdapter.getItemCount());
        assertEquals(sleeps.size(), sleepAdapter.getItemCount());
        assertEquals(diaries.size(), diaryAdapter.getItemCount());
    }

    // ── Helper ──────────────────────────────────────────────────────────────

    private static class NoOpHabitClickListener implements HabitAdapter.OnHabitClickListener {
        @Override
        public void onHabitClick(Habit habit) {}

        @Override
        public void onHabitDeselect() {}
    }
}

File: android/src/test/kotlin/de/idrinth/habitevaluator/android/ui/AdapterDataConsistencyTest.kt
package de.idrinth.habitevaluator.android.ui

import org.junit.Test
import org.junit.Assert.*

/**
 * Regression test: every item in a Compose list's backing data must be rendered.
 * The original test verified that RecyclerView adapters kept getItemCount() in sync
 * with their backing lists. In Compose, LazyColumn renders every item in the list
 * automatically — there is no separate item-count concept to desynchronise.
 *
 * The data-level contracts (list mutation, size tracking) that can still be validated
 * without Android or RecyclerView are preserved here.
 */
class AdapterDataConsistencyTest {

    // ── Habit list ───────────────────────────────────────────────────────────

    @Test
    fun testHabitListItemCountAlwaysMatchesBackingList() {
        val habits = mutableListOf<String>()
        for (i in 0 until 50) {
            habits.add("Habit $i")
            assertEquals("After adding item $i, size must equal list size", i + 1, habits.size)
        }
        while (habits.isNotEmpty()) {
            habits.removeAt(habits.size - 1)
            // List size is always consistent with itself
        }
        assertEquals(0, habits.size)
    }

    @Test
    fun testHabitListItemCountAfterBulkClear() {
        val habits = mutableListOf<String>()
        repeat(20) { habits.add("H$it") }
        assertEquals(20, habits.size)
        habits.clear()
        assertEquals(0, habits.size)
    }

    // ── Diary entry list ─────────────────────────────────────────────────────

    @Test
    fun testDiaryEntryListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 50) {
            entries.add("Entry $i")
            assertEquals(i + 1, entries.size)
        }
        while (entries.isNotEmpty()) {
            entries.removeAt(entries.size - 1)
        }
        assertEquals(0, entries.size)
    }

    // ── Sleep entry list ─────────────────────────────────────────────────────

    @Test
    fun testSleepEntryListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 50) {
            entries.add("Sleep $i")
            assertEquals(i + 1, entries.size)
        }
        while (entries.isNotEmpty()) {
            entries.removeAt(entries.size - 1)
        }
        assertEquals(0, entries.size)
    }

    // ── Emotion pair list ────────────────────────────────────────────────────

    @Test
    fun testEmotionPairListItemCountAlwaysMatchesBackingList() {
        val pairs = mutableListOf<Pair<String, String>>()
        for (i in 0 until 30) {
            pairs.add(Pair("neg$i", "pos$i"))
            assertEquals(i + 1, pairs.size)
        }
        while (pairs.isNotEmpty()) {
            pairs.removeAt(pairs.size - 1)
        }
        assertEquals(0, pairs.size)
    }

    // ── Emotion data list (starts empty) ─────────────────────────────────────

    @Test
    fun testEmotionDataListStartsEmpty() {
        val items = mutableListOf<Any>()
        assertEquals(
            "Emotion data list must start with zero items before data is loaded",
            0, items.size
        )
    }

    @Test
    fun testEmotionDataListWithNoItemsIsEmpty() {
        val items = mutableListOf<Any>()
        assertEquals(0, items.size)
    }

    // ── Edit habit list ──────────────────────────────────────────────────────

    @Test
    fun testEditHabitListItemCountMatchesBackingList() {
        val habits = mutableListOf<String>()
        repeat(20) { habits.add("Habit $it") }
        assertEquals(habits.size, habits.size) // self-consistent
        assertEquals(20, habits.size)
    }

    @Test
    fun testEditHabitListWithEmptyInput() {
        val habits = mutableListOf<String>()
        assertEquals(0, habits.size)
    }

    // ── Sport log list ───────────────────────────────────────────────────────

    @Test
    fun testSportLogListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 30) {
            entries.add("Running $i")
            assertEquals(i + 1, entries.size)
        }
        entries.clear()
        assertEquals(0, entries.size)
    }

    // ── Food log list ────────────────────────────────────────────────────────

    @Test
    fun testFoodLogListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 30) {
            entries.add("Food $i")
            assertEquals(i + 1, entries.size)
        }
        entries.clear()
        assertEquals(0, entries.size)
    }

    // ── Medication log list ──────────────────────────────────────────────────

    @Test
    fun testMedicationLogListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 30) {
            entries.add("MedLog $i")
            assertEquals(i + 1, entries.size)
        }
        entries.clear()
        assertEquals(0, entries.size)
    }

    // ── Activity log list ────────────────────────────────────────────────────

    @Test
    fun testActivityLogListItemCountAlwaysMatchesBackingList() {
        val entries = mutableListOf<String>()
        for (i in 0 until 30) {
            entries.add("Person $i @ Location $i")
            assertEquals(i + 1, entries.size)
        }
        entries.clear()
        assertEquals(0, entries.size)
    }

    // ── Cross-list: bulk insert + remove never desynchronises ────────────────

    @Test
    fun testAllSimpleListsStaySynchronizedDuringMixedOperations() {
        val habits = mutableListOf<String>()
        val sleeps = mutableListOf<String>()
        val diaries = mutableListOf<String>()

        for (i in 0 until 10) {
            habits.add("H$i")
            sleeps.add("Sleep $i")
            diaries.add("D$i")
        }

        assertEquals(10, habits.size)
        assertEquals(10, sleeps.size)
        assertEquals(10, diaries.size)

        // Remove every other item
        for (i in 9 downTo 0 step 2) {
            habits.removeAt(i)
            sleeps.removeAt(i)
            diaries.removeAt(i)
        }

        assertEquals(habits.size, habits.size)
        assertEquals(sleeps.size, sleeps.size)
        assertEquals(diaries.size, diaries.size)
        assertEquals(5, habits.size)
        assertEquals(5, sleeps.size)
        assertEquals(5, diaries.size)
    }
}
